### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by FlamingLeo**
Finished worksheet 4.
- Added thermostat.
- Added periodic boundaries.
- Added checkpointing.
- Added custom timing library, removed `gbench`.
- Added profiling options.
- `sigma` and `epsilon` are now particle-specific.
- Removed raw text file input.
- Various minor additions, removals and improvements.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -172,14 +164,31 @@ Complete XML input files have the following structure:
...
+    <!-- (Optional) The target temperature of the system. -->
+    <!-- If the target temperature is not specified, the initial temperature is used. -->
+    <target><!-- double --></target>
+    <!-- The maximum temperature difference in one thermostat application. -->
+    <!-- If it is not specified, the temperature will be updated directly. -->
+    <!-- Note that this may cause abrupt changes in the particle velocities. -->
+    <deltaT><!-- double --></deltaT>
+    <!-- Specify, whether or not particle velocities should be initialized with Brownian Motion in the first iteration. -->
+    <brownianMotion><!-- boolean --></brownianMotion>
+  </thermostat>
```
Comment (SamNewcome) [inline README.md]: I would add a short comment explaining what to do if you do not want to use the thermostat.

File: tests/objects/TestThermostat.cpp
Comment (SamNewcome) [inline tests/objects/TestThermostat.cpp:1]: Nice! But add a version of these tests with `dim=3` for WS5.

File: src/strategies/BoundaryConditions.h
Comment (SamNewcome) [inline src/strategies/BoundaryConditions.h:1]: The name `targetCell` is a little confusing. This probably made sense when Manish reviewed it in WS3, because he knew the full context. But I jumped to `handlePeriodicCondition` because this was the new feature and was confused by it because there it means the cell the particle has moved to but not necessarily the cell it will end up in (which could also be called the "target" cell in that context). Potentially just adding to the documentation for it will suffice as opposed to changing the name.

File: src/strategies/BoundaryConditions.cpp
Comment (SamNewcome) [inline src/strategies/BoundaryConditions.cpp:1]: The logic is confusing but looks correct. The primary cause of confusion is the handling of recursion and boundary condition mixing.

You might want to consider alternatively looping through all cardinal directions and, for each, checking if the target cell is along that boundary (including edge/corner cells), checking what kind of boundary it is, applying that boundary, and then moving on to the next cardinal direction. This would avoid the complicated handling of corners/edges (which might make 3D easier and cleaner) and removes the need for recursion. This, of course, is not an insignificant amount of work, so you should consider if this is worth it or not.

File: tests/strategies/TestBoundaryConditions.cpp
Comment (SamNewcome) [inline tests/strategies/TestBoundaryConditions.cpp:1]: A mixed periodic/reflective corner case test would be good.

File: src/strategies/PositionCalculation.h [hunk 2]
```diff
@@ -38,6 +39,7 @@ void calculateX(ParticleContainer &particles, double delta_t, CellContainer *lc
  *
  * @param particles The ParticleContainer containing the Particle objects to iterate over.
  * @param delta_t The timestep \f$ \Delta t \f$.
+ * @param g_grav The gravitational force \f$ g_{grav} \f$.
  * @param lc The CellContainer for the linked cells method.
  */
-void calculateX_LC(ParticleContainer &particles, double delta_t, CellContainer *lc);
```
Comment (SamNewcome) [inline src/strategies/PositionCalculation.h:43]: Whilst you have added documentation for `g_grav`, you have not added to the brief or details anything to do we adding gravity. 

It would be even cleaner to give the function a name with `calculateX_LC_with_gravity` or `calculateX_LC_and_reset_force` to make it clearer.

There is also no unit test for gravity.

File: src/strategies/ForceCalculation.cpp [hunk 3]
```diff
@@ -1,23 +1,50 @@
...
-    for (auto &p1 : particles) {
-        for (auto &p2 : particles) {
-            if (p1 != p2) {
-                p1.setF(p1.getF() + ArrayUtils::elementWiseScalarOp(
-                                        p1.getM() * p2.getM() / std::pow(ArrayUtils::L2Norm(p1.getX() - p2.getX()), 3),
-                                        p2.getX() - p1.getX(), std::multiplies<>()));
-            }
-        }
+// helper function to calculate force vector
+std::array<double, 3> getForceVec(const Particle &p1, const Particle &p2, const std::array<double, 3> distVec,
```
Comment (SamNewcome) [inline src/strategies/ForceCalculation.cpp]: This is a confusing name because it is only relevant for LJ not the pairwise WS1 gravity. But it is nicer that you have this as a helper function.

File: tests/strategies/TestForceCalculation.cpp
Comment (SamNewcome) [inline tests/strategies/TestForceCalculation.cpp:1]: No test for particle mixing.

File: src/io/output/XMLWriter.cpp
Comment (SamNewcome) [inline src/io/output/XMLWriter.cpp:1]: I don't see a test for the checkpointing/XMLWriting

**Review by SamNewcome — APPROVED** [review]
Nice!

Particularly the optimisations were really nice.

Within the code itself, the main issue is some missing unit tests and that the code for boundaries is a little confusing (but only a little).

File: src/strategies/BoundaryConditions.cpp
Comment (FlamingLeo) [reply src/strategies/BoundaryConditions.cpp:1]: 3D should already be finished, so we probably won't use this, but it's a nice idea nevertheless. All the checks mentioned might negatively affect performance however, and we haven't had any issues with recursion yet, even for the largest examples.

File: src/strategies/ForceCalculation.cpp (hunk 3 above)
Comment (FlamingLeo) [reply src/strategies/ForceCalculation.cpp]: Fair - we didn't put that much thought into the name since it's not meant to be a publicly visible function. Rather, it's an undocumented internal helper function of the you-know-what-you're-doing kind. But I've changed the name now and made the functions `static` to better reflect our intentions.

File: src/strategies/PositionCalculation.h (hunk 2 above)
Comment (FlamingLeo) [reply src/strategies/PositionCalculation.h:43]: This was a late optimization, so I probably just forgot to describe what happens to the force at the end of the function. I've now added a description in the documentation.

As for changing the name of the functions, I might do this later, because I fear changing the names of all occurences now could lead to some merge conflicts.

Unit tests coming up.
